samples/cellular/modem_shell/src/sock/sock_shell.c

```diff
@@ -852,11 +931,248 @@ static int cmd_sock_rai(const struct shell *shell, size_t argc, char **argv)
+	}
+
+	/* Variables for command line arguments */
+	int arg_socket_id = SOCK_ID_NONE;
+	int arg_sock_opt_level = SOL_SOCKET;
+	int arg_sock_opt_id = 0;
+	char *arg_sock_opt_value = NULL;
+	struct sock_opt_entry *sock_opt;
+	struct sock_level_entry *sock_level;
+
+	optreset = 1;
+	optind = 1;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
+
+		switch (opt) {
+		case 'i': /* Socket ID */
+			arg_socket_id = atoi(optarg);
+			break;
+
+		case SOCK_SHELL_OPT_SOCKET_OPTION_LEVEL: /* Level */
+			arg_sock_opt_level = atoi(optarg);
+			if (arg_sock_opt_level == 0) {
+				sock_level = sock_level_map(optarg);
+				if (sock_level == NULL) {
+					mosh_error("Unknown socket level: %s", optarg);
+					goto show_usage;
+				}
+				arg_sock_opt_level = sock_level->level;
+			}
+			break;
+
+		case 'o': /* Option */
+			arg_sock_opt_id = atoi(optarg);
+			if (arg_sock_opt_id == 0) {
+				sock_opt = sock_opt_map(optarg);
+				if (sock_opt == NULL) {
+					mosh_error("Unknown socket option: %s", optarg);
+					goto show_usage;
+				}
+				arg_sock_opt_id = sock_opt->opt;
+				arg_sock_opt_level = sock_opt->level;
+			}
+			break;
+
+		case 'v': /* Value */
+			arg_sock_opt_value = optarg;
+			break;
+
+		case 'h':
+			goto show_usage;
+		case '?':
+		default:
+			mosh_error("Unknown option (%s). See usage:", argv[optind - 1]);
+			goto show_usage;
+		}
+	}
+
+	if (optind < argc) {
+		mosh_error("Arguments without '-' not supported: %s", argv[argc - 1]);
+		goto show_usage;
+	}
+
+	err = sock_setopt(arg_socket_id, arg_sock_opt_level, arg_sock_opt_id, arg_sock_opt_value);
+
+	return err;
+
+show_usage:
+	sock_print_usage(SOCK_CMD_OPTION_SET);
+	return err;
+}
+
+static int cmd_sock_getopt(const struct shell *shell, size_t argc, char **argv)
+{
+	int err = 0;
+
+	if (argc < 2) {
+		goto show_usage;
+	}
+
+	/* Variables for command line arguments */
+	int arg_socket_id = SOCK_ID_NONE;
+	int arg_sock_opt_level = SOL_SOCKET;
+	int arg_sock_opt_id = -1;
+	struct sock_opt_entry *sock_opt;
+	struct sock_level_entry *sock_level;
+
+	optreset = 1;
+	optind = 1;
+	int opt;
+
+	while ((opt = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
+
+		switch (opt) {
+		case 'i': /* Socket ID */
+			arg_socket_id = atoi(optarg);
+			break;
+
+		case SOCK_SHELL_OPT_SOCKET_OPTION_LEVEL: /* Level */
+			arg_sock_opt_level = atoi(optarg);
+			if (arg_sock_opt_level == 0) {
+				sock_level = sock_level_map(optarg);
+				if (sock_level == NULL) {
+					mosh_error("Unknown socket level: %s", optarg);
+					goto show_usage;
+				}
+				arg_sock_opt_level = sock_level->level;
+			}
+			break;
+
+		case 'o': /* Option */
+			arg_sock_opt_id = atoi(optarg);
+			if (arg_sock_opt_id == 0) {
+				sock_opt = sock_opt_map(optarg);
+				if (sock_opt == NULL) {
+					mosh_error("Unknown socket option: %s", optarg);
+					goto show_usage;
+				}
+				arg_sock_opt_id = sock_opt->opt;
+				arg_sock_opt_level = sock_opt->level;
```

Comment:
Does this override level if you have first `--level` and then `-o`?

Reply:
Yes, it does. If the user gives an incorrect level with `--level` and specifies the socket option using textual presentation with `-o`, the incorrect level would be ignored and the correct one from the mapping table used. If the order of the options is reversed, the incorrect level would be used and an error would be returned from the modem.

I did consider this myself, but I decided not to add a check here so that we could return an error the user instead of executing the command successfully. :)

Reply:
I think this should behave similarly irrespective of the order. We haven't broken that "rule" to my knowloedge, except when we clearly say to what the order impacts, i.e., `-m` options in `location get`.

My intention with `--level` was that if it's not set, `-o` automatically sets it. However, if `--level` is given, user specifically wants to try certain level value for a given option and we respect that choice. This also enables testing of some incorrect combinations so that libmodem and modem properly responds to those.

That being said, error combinations could be verified with integer values for the option so above is negotiable. However, the ordering issue is another thing and I think it should be made transparent to the ordering of parameters.